lib/pengine/pe_actions.c

```diff
@@ -74,71 +74,112 @@ find_existing_action(const char *key, const pcmk_resource_t *rsc,
+        const char *interval_spec = crm_element_value(operation,
+                                                      XML_LRM_ATTR_INTERVAL);
 
-            name = crm_element_value(operation, "name");
-            interval_spec = crm_element_value(operation, XML_LRM_ATTR_INTERVAL);
-            if (!include_disabled && pcmk__xe_get_bool_attr(operation, "enabled", &enabled) == pcmk_rc_ok &&
-                !enabled) {
-                continue;
-            }
+        // @TODO This does not consider rules, defaults, etc.
+        if (!include_disabled
+            && (pcmk__xe_get_bool_attr(operation, "enabled",
+                                       &enabled) == pcmk_rc_ok) && !enabled) {
+            continue;
+        }
 
-            interval_ms = crm_parse_interval_spec(interval_spec);
-            match_key = pcmk__op_key(rsc->id, name, interval_ms);
-            if (pcmk__str_eq(key, match_key, pcmk__str_casei)) {
-                op = operation;
-            }
-            free(match_key);
+        interval_ms = crm_parse_interval_spec(interval_spec);
 
-            if (rsc->clone_name) {
-                match_key = pcmk__op_key(rsc->clone_name, name, interval_ms);
-                if (pcmk__str_eq(key, match_key, pcmk__str_casei)) {
-                    op = operation;
-                }
-                free(match_key);
-            }
+        // Try first with resource ID
+        match_key = pcmk__op_key(rsc->id, name, interval_ms);
+        if (pcmk__str_eq(key, match_key, pcmk__str_none)) {
+            action_config = operation;
+        }
+        free(match_key);
 
-            if (op != NULL) {
-                free(local_key);
-                return op;
+        // Then again with clone_name in case ID has instance number
+        if (rsc->clone_name != NULL) {
+            match_key = pcmk__op_key(rsc->clone_name, name, interval_ms);
+            if (pcmk__str_eq(key, match_key, pcmk__str_none)) {
+                action_config = operation;
             }
+            free(match_key);
+        }
+
+        if (action_config != NULL) {
+            break;
         }
     }
 
-    free(local_key);
-    if (do_retry == FALSE) {
-        return NULL;
+    return action_config;
+}
+
+/*!
+ * \internal
+ * \brief Find the XML configuration of a resource action
+ *
+ * \param[in] rsc               Resource to find action configuration for
+ * \param[in] key               "RSC_ACTION_INTERVAL" of action to find
+ * \param[in] include_disabled  If false, do not return disabled actions
+ *
+ * \return XML configuration of desired action if any, otherwise NULL
+ */
+static xmlNode *
+find_rsc_op_entry_helper(const pcmk_resource_t *rsc, const char *key,
+                         bool include_disabled)
+{
+    char *retry_key = NULL;
+    xmlNode *action_config = NULL;
+
+    // Try exact key first
+    action_config = find_exact_action_config(rsc, key, include_disabled);
+    if (action_config != NULL) {
```

Comment:
why not 

```
while (action_config == NULL) {
  if {compare migrate} {
    retry_key = "migrate";
  } else if {compare notify} {
    retry_key = "notify";
  } else break;
  // or test for retry_key here - you know I like breaking out of blocks ;-)
  action_config = find_exact ...
}
return action_config;
```

it is just a few lines so no need to trade lower block-depth against obfuscating actual logic.
Again haven't checked yet if that goes away later yet.